### ISSUE TITLE
Random `Target of URI doesn't exist: 'package:foobar/foobar.dart'. Try creating the file referenced by the URI, or try using a URI for a file that does exist` red errors scattered everywhere in the source code

### ISSUE THREAD

**Issue by andynewman10** [Dart-Code/Dart-Code#5171]
**Describe the bug**

VS code sometimes shows plenty of red errors in Explorer view and in the source code, but it appears to me that there are indeed no errors in my source code.

**To Reproduce**

It is difficult to reproduce and it happens usually with large projects.

This is what I usually do:

1. Open VS code, which opens my trusted workspace folder
2. In my flutter application folder, run `flutter clean` from the command line. Sometimes, I also do this in flutter project sub-folders that are referenced from my master app.
3. Open the command palette and reload the VS window (i.e. restart VS code)
4. Let the 'Run flutter pub get?' popup appear, and run the command
5. Let VS code cool down until code analysis has completely finished its job. I have a large project, so I usually leave it plenty of time as I know code analysis will take a long time.

**Expected behavior**

The above mentionned steps, for a project that is supposed to have no compilation error, should result in a display with no error (no red text).

**Actual behavior**

Following the above mentionned steps still results in red errors scattered everywhere in the project and sub-projects (located in subfolders) in my workspace.

Hovering over an import package declaration displays a tooltip window `Target of URI doesn't exist: 'package:foobar/foobar.dart'. Try creating the file referenced by the URI, or try using a URI for a file that does exist.`

**Workaround**

The workaround is extremely annoying.

As I said, it seems that there are no errors in the source code. To force VS code to somehow "redo" its code analysis, I edit the pubspec.yaml file, add a blank space somewhere, and save it. Code analysis kicks in and... there are no more errors!

Note that I have to do that at multiple places, i.e. for most if not all pubspec.yaml files in sub-projects. By sub-projects, I mean projects declared that way in my application's pubspec.yaml file:

```
foobarsubproject:
  path: ./foobarsubproject
```
Is there any Dart Dev Tools command that I could use to automate that, or just "refresh" code analysis?

**Please complete the following information:**

 - Operating System and version: Windows 11 21H2
 - VS Code version: 1.90.2
 - Dart extension version: 3.92.0
 - Dart/Flutter SDK version: 3.22.2
 - Target device (if the issue relates to Flutter debugging): Web

**Comment by DanTup** [comment]
Can you give more details about where the errors occur and which project they point to? It's not clear to me if by:

```yaml
foobarsubproject:
  path: ./foobarsubproject
```

You mean you are adding a reference to "itself" here, or if you just mean you're using `path` dependencies between local projects. When this happens, if you open up the `.dart_tool/package_config.json` for the project where the analysis error is shown, does it contain a reference to the package that is "not found" and the correct path to it?

Does opening a file that has this error cause the error to go away?

Do you know when this issue started? Is it with a recent update of Dart/Flutter?

**Comment by andynewman10** [comment]
My project structure is the following:
```
+ workspaceFolder (the folder I open in VS code)
--+ flutterapp
-----+ pubspec.yaml (and .lock)
-----+ subproject1
---------+ pubspec.yaml (and .lock)
-----+ subproject2
---------+ pubspec.yaml (and .lock)
```
The pubspec.yaml for flutterapp looks like:
```
dependencies:
  subproject1:
    path: ./subproject1
  subproject2:
    path: ./subproject2
```

> When this happens, if you open up the .dart_tool/package_config.json for the project where the analysis error is shown, does it contain a reference to the package that is "not found" and the correct path to it?

`subproject1`, for instance, has red errors, among which one of the errors ('Target URI...') is on this line:
```
import 'package:google_maps/google_maps.dart' as gmaps;
```
Its pubspec.yaml has a dependency on `google_maps`, and I am seeing a `google_maps` entry in the `subfolder1/.dart_tool/package_config.json` file.

It is difficult to say if this issue is recent (Flutter 3.22.2) or not.

**Comment by DanTup** [comment]
Thanks - I don't suppose this is a public project (or you're able to reproduce the issue in a public project)? Being able to test with a project known to trigger this (or get logs that won't include anything sensitive) would really help.

I was going to write instructions for checking the package resolution data in the analyzer when you next see this, but unfortunately it looks like there is an issue with this data not showing up in the diagnostic pages - I'll work on a fix for that (but unfortunately it'll be a while before that makes a stable SDK).

**Comment by andynewman10** [comment]
Unfortunately it is not a public project.

To be on the safe side, I tried on another Windows computer (same Flutter and Dev Tools versions) and got the same result. Not with my Mac right now, but I suspect the issue indeed can be observed on my Mac setup as well (I didn't do a 'flutter clean' this afternoon on that machine, which might be why it was working).

What is weird is that not all packages seem to be affected by the issue. There seems to be a pattern that I have yet to understand, or at least try to.

I just don't understand why simply forcing a save (Ctrl+S) of a pubspec.yaml file would help get rid of the errors. If you have some magic command to help diagnose the issue of something to try, I'll surely give it a go.

As long as there are red errors in VS code, Dev Tools won't let me start my debug session, which is quite severe.

**Comment by DanTup** [comment]
> What is weird is that not all packages seem to be affected by the issue. There seems to be a pattern that I have yet to understand, or at least try to.

Out of interest, could it be that those that are also dependencies of the parent project work? [An issue](https://github.com/dart-lang/sdk/issues/56047) came up recently where the workspace root was being used for resolution inside analysis_options files where a nested projects package config should've been used. I don't think it's the same issue (I think that issue is specific to analysis_options, and only in `master`/`beta` but I wonder if it could be something similar.

If you can reproduce this fairly reliably, I wonder if you could copy the project and strip out almost all of the code to see if you can make a reproducible sample that would be shareable (or that you could capture logs for without anything sensitive being included)?

**Comment by andynewman10** [comment]
Interesting.
I am trying to narrow is down to something more manageable. Let's take this example:
```
+ workspaceFolder (the folder I open in VS code)
--+ flutterapp
-----+ pubspec.yaml (and .lock)
-----+ my_google_maps
---------+ pubspec.yaml (and .lock)
-----+ my_google_maps_flutter_web
---------+ pubspec.yaml (and .lock)
```
`my_google_maps` and `my_google_maps_flutter_web` are local, customized versions of `google_maps` and `google_maps_flutter_web`.

The pubspec.yaml for flutterapp has a dependency_override for `google_maps`:
```
dependency_overrides:
  google_maps:
    path: ./my_google_maps
```
The pubspec.yaml for `my_google_maps_flutter_web` is like this:
```
dependencies:
  google_maps: ^7.1.0
```
I would like my top-level (app-level) dependency_override for `google_maps` to be global, that is, I expect all subprojects, including `my_google_maps_flutter_web`, should use it, and never use `google_maps`. But it seems like it is not the case. What is the expected behaviour for this? How does transitive dependencies work in this case - can an application be built with multiple package versions, selectively used by one particular package or another?

After a `flutter clean` at app level (which, BTW, I expect to be a flutter clean for all subprojects as well - is an app-level flutter clean sufficient?), then a subsequent app-level `flutter pub get` results in:
- flutterapp's pubspec.lock contains my custom google_maps, so does package_config.json
- my_google_maps_flutter_web's pubspec.lock contains the public google_maps, not my own version's, so does package_config.json

**Comment by andynewman10** [comment]
- I cannot reproduce the problem on my Mac today.
- On my first Windows 11 PC, flutter clean, window reload, then "Some packages are out of date... run flutter pub get?" -> yes. 'Analyzing...' runs for a very long time, then disappears. Almost everything stays red. Window reload -> red errors all disappear (I have observed this behaviour already a long time ago).
- On my second Windows 11 PC, flutter clean, window reload, then "Some packages are out of date... run flutter pub get?" -> yes. 'Analyzing...' runs for some time, then disappears. Many things stay red. Window reload -> no effect, stays the same.

All three computers are from ~2016 with comparable performance (Skylake processors). By today's standards, I would say they have low performance.

The only difference between my Mac and my PCs is that on my PCs, yesterday, I issue a 'flutter pub clean cache' at some point because I am importing a dependency from a git branch and I wanted to be sure the latest commit would be picked up.

**Comment by andynewman10** [comment]
On my second Windows 11 PC, a window reload actually removed all, but one error: the google map error I was talking about. This remaining red error is in the my_google_maps_flutter_web subproject. And then, if I open this subproject's pubspec.yaml file, just hit Ctrl+S, 'Analyzing...' kicks in and the problem disappears.

I think the thing I should do at this point is to make sure my source tree is Dev-Tools files free.

How can I make sure I purge all dev-tools related files, apart from issuing flutter clean everywhere I can?

**Comment by DanTup** [comment]
> What is the expected behaviour for this? How does transitive dependencies work in this case - can an application be built with multiple package versions, selectively used by one particular package or another?

For analysis, the analyzer will always resolve packages based on what's in the `.dart_tool/package_config.json` file for that project. Using a `dependency_override` in your top-level app will work when running the app (because there will be a single solve for the whole app, which will use the overrides), but it won't affect analysis of the child projects in VS Code, because those child projects could be referenced by other apps that do not provide these overrides.

So, you should add dependency_overrides to each of your packages if you want them to be analyzed using the overrides (note: there has been discussion of better "monorepo" support where you could define that an entire tree has a single solve, but I don't think that's available yet).


> The only difference between my Mac and my PCs is that on my PCs, yesterday, I issued a 'flutter pub clean cache' at some point

Deleting things from your pubcache can cause issues if you don't re-run `pub get` in projects, because they might have paths (in `.dart_tool/package_config.json`) to missing folders. In order to eliminate this as a source of the problem, I would manually run `flutter pub get` in each of the projects individually (this will restore any packages being pulled from pub in the pubcache) and see if you can reproduce.


> How can I make sure I purge all dev-tools related files, apart from issuing flutter clean everywhere I can?

`flutter clean` will delete the `.dart_tool` folders which includes all of the package resolution info. You'd need to do this for each project though. You can also manually search for `.dart_tool` folders and delete all of those. Are you doing this for testing, or do you think you're close to having a shareable project that reproduces the issue? If the latter, you'd need to also be sure there aren't things like `build` folders that could have compiled output in them.

**Comment by andynewman10** [comment]
> > What is the expected behaviour for this? How does transitive dependencies work in this case - can an application be built with multiple package versions, selectively used by one particular package or another?
> 
> For analysis, the analyzer will always resolve packages based on what's in the `.dart_tool/package_config.json` file for that project. Using a `dependency_override` in your top-level app will work when running the app (because there will be a single solve for the whole app, which will use the overrides), but it won't affect analysis of the child projects in VS Code, because those child projects could be referenced by other apps that do not provide these overrides.
> 
> So, you should add dependency_overrides to each of your packages if you want them to be analyzed using the overrides (note: there has been discussion of better "monorepo" support where you could define that an entire tree has a single solve, but I don't think that's available yet).

Thanks. That is some important piece of information I was unaware of! The analyzer might indeed show errors that do not exist while building the app. Wow.
Yes, it would be interesting to have some kind of optional context in Dev Tools where it sees things through a "master" project. Difficult decision obviously.
> 
> > The only difference between my Mac and my PCs is that on my PCs, yesterday, I issued a 'flutter pub clean cache' at some point
> 
> Deleting things from your pubcache can cause issues if you don't re-run `pub get` in projects, because they might have paths (in `.dart_tool/package_config.json`) to missing folders. In order to eliminate this as a source of the problem, I would manually run `flutter pub get` in each of the projects individually (this will restore any packages being pulled from pub in the pubcache) and see if you can reproduce.
> 
> > How can I make sure I purge all dev-tools related files, apart from issuing flutter clean everywhere I can?
> 
> `flutter clean` will delete the `.dart_tool` folders which includes all of the package resolution info. You'd need to do this for each project though. You can also manually search for `.dart_tool` folders and delete all of those. Are you doing this for testing, or do you think you're close to having a shareable project that reproduces the issue? If the latter, you'd need to also be sure there aren't things like `build` folders that could have compiled output in them.

Makes sense. I re-ran pub get on each project and subproject recursively to be on the safe side.

I do believe there is a bug somewhere with Dev Tools and/or VS code.

1. open VS code
2. flutter clean, then Window: reload immediately
3. 'Some packages are out of date..., would you like to run pub get?' -> yes
4. Red everywhere and 'Analyzing...' shows up in the bottom bar
5. Nothing shows up anymore in the bottom bar for about 20 seconds. Possibly because I have a large workspace.
6. The Errors|Warnings|Infos counters increase at the extreme bottom left corner
7. The Run|Debug|Profile indicator above the main() function in my opened Dart file shows up.
8. At this point things are stable. Red errors all over the place
9. Window: reload -> no red errors at all.

**Comment by DanTup** [comment]
> Thanks. That is some important piece of information I was unaware of! The analyzer might indeed show errors that do not exist while building the app. Wow.
> Yes, it would be interesting to have some kind of optional context in Dev Tools where it sees things through a "master" project. Difficult decision obviously.

I think doing it that way would just lead to similar issues in the opposite direction. For example you don't see analysis errors in ProjectB because it's using overrides from ProjectA, then when you try to run tests in ProjectB, they fail because it's not using the override (because running tests in ProjectB has nothing to do with ProjectA).

I think the current behaviour is probably the best, but it should perhaps be clearer that dependency_overrides are only in the context of the application that defines them, they're not inherited or otherwise applied to other projects. The analyzer always analyzes each project in its own context (at least until the monorepo support noted above may be implemented).

> I do believe there is a bug somewhere with Dev Tools and/or VS code.

Are you able to make a project that triggers this that I can test with? Possibly copying your project and deleting all of the code (just leaving some imports to illustrate the problem) would be enough. It'll be hard to debug without a repro (or at least a log) and I think you'll have more success creating one than I, since you have a known project that's triggering this.

Thanks!

**Comment by andynewman10** [comment]
I am very sorry for my late reply. So far I have been unable to understand what's going on, unfortunately too busy with something else. I suspect I strongly suspect I have had this problem in the past, months ago if not more.

> I think doing it that way would just lead to similar issues in the opposite direction. For example you don't see analysis errors in ProjectB because it's using overrides from ProjectA, then when you try to run tests in ProjectB, they fail because it's not using the override (because running tests in ProjectB has nothing to do with ProjectA).

Correct, but in this case the developer would be deliberately conscious that his/her global context prevails and that Dart code analysis is deliberately performed within this global context (global context = dependency override context). CMake tools have this kind of concept, since you must tell CMake tools which CMakeLists.txt governs everything (CMake tools demands you to select the CMakeLists.txt file, and the compiler ; both represent the global context).

**Comment by DanTup** [comment]
> Correct, but in this case the developer would be deliberately conscious that his/her global context

But this isn't "global context", this is context from "one of the other projects in this workspace".

Imagine your workspace looks like:

```
- projectA
- projectB
- sharedLib
```

It might be that `projectA` and `projectB` both resolve some dependencies to different versions because they are constrained by their own (different) dependencies (or overrides).

The monorepo work I mentioned above _is_ about having some global context, and that may address this. Its purpose is to lock a pub solve for the whole workspace/monorepo, so all of the packages are using the same versions of everything (I think it will work by essentially having a single `.dart_tool/package_config.json` at the root, rather than one for each project).

**Comment by andynewman10** [comment]
Or, you could leave things as they are, with a `.dart_tool/package_config.json` per project. Instead, Dev Tools could have a setting where you specify which `build.yaml` (and dependency_override section) rules things workspace-wide for dependency analysis, a little bit like CMake tools has a setting somewhere in memory telling which `CMakeLists.txt` file rules (you can easily reset the setting with `CMake: Configure`).

**Comment by DanTup** [comment]
> Or, you could leave things as they are, with a `.dart_tool/package_config.json` per project. Instead, Dev Tools could have a setting where you specify which `build.yaml` (and `dependency_override` section) rules things workspace-wide

The purpose of `.dart_tool/package_config.json` is to provide the mapping of where packages will be loaded from. I think unlikely (and I don't think it makes sense) that all of the existing Dart tooling (including the analyzer and `dart run`) would be changed to have another parallel way to define this. The correct way to do this is to ensure the `.dart_tool/package_config.json` being used contains the actual packages you want to use.

The current way to do this is to ensure you have set the overrides in the packages, but in future there will be work to support a monorepo where the resolution is the same across the whole repository. It's not clear to me why there needs to be another way of achieving the same thing.



> I have at the moment VS code open on my PC (exhibiting the bug) and my Mac (working).
>
> - on my PC, plenty of red errors. It stays like that. If I do Developer: Reload window, all red errors are gone in this file.

Can you reliably reproduce this? Does it only happen in this specific workspace, or can you reproduce in a new workspace (or something you can share so I can do the same steps)? An instrumentation log may be useful, but being able to reproduce it here would be much better because I could more easily debug. Thanks!

**Comment by andynewman10** [comment]
> > I have at the moment VS code open on my PC (exhibiting the bug) and my Mac (working).
> > 
> > * on my PC, plenty of red errors. It stays like that. If I do Developer: Reload window, all red errors are gone in this file.
> 
> Can you reliably reproduce this? Does it only happen in this specific workspace, or can you reproduce in a new workspace (or something you can share so I can do the same steps)? An instrumentation log may be useful, but being able to reproduce it here would be much better because I could more easily debug. Thanks!

I can reproduce it reliably with my main workspace. If I create another workspace, I can't repro.

**Comment by DanTup** [comment]
If you copy your `pubspecs`/`dependency_overrides` into the new workspace so that the layout is similar, can you reproduce it? If not, if you completely duplicate your workspace, does it repro there? (If so, is it possible you could try to trim it down until you either figure out what the trigger is, or you have something that has nothing private remaining so that you could share the repro)? Thanks!

**Comment by andynewman10** [comment]
Copying over my `.vscode` folder to another blank workspace and trying to reproduce the problem in that workspace doesn't work. It does not seem immediately related to the `.vscode` folder. Starting to chase what's going on in my main workspace is challenging. I can try to duplicate it and see how VS behaves.

**Comment by DanTup** [comment]
Thanks - if you're able to get to a repro you can share, that would help _a lot_. It'll be difficult for me to do without a project that triggers it to use as a reference, and it may also be difficult to get detailed enough logs to understand what's happening without them including parts of your project.

By the way - are you reproducing this on a _stable_ branch (and the new project you're creating with `flutter create` is using a stable version of Flutter from `PATH`)? (I want to exclude https://github.com/dart-lang/sdk/issues/56047 from being related, but if you're on beta/master, there's a chance that's playing a part here).

**Comment by andynewman10** [comment]
Sharing something or deep-analysing my workspace is too challenging right now.

I am on the stable branch of Flutter, version 3.22.2.

I disabled the Microsoft C/C++ and Java extension packs, then reloaded the window, to be on the safe side. Same result.

With VS code open and just sitting, if I launch `flutter create foobarapp4 --platforms web` from a separate terminal, VS code reacts, then I get the following in package_config.json:
```
{
  "configVersion": 2,
  "packages": [
    {
      "name": "async",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/async-2.11.0",
      "packageUri": "lib/",
      "languageVersion": "2.18"
    },
    {
      "name": "boolean_selector",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/boolean_selector-2.1.1",
      "packageUri": "lib/",
      "languageVersion": "2.17"
    },
    {
      "name": "characters",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/characters-1.3.0",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "clock",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/clock-1.1.1",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "collection",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/collection-1.18.0",
      "packageUri": "lib/",
      "languageVersion": "2.18"
    },
    {
      "name": "cupertino_icons",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/cupertino_icons-1.0.8",
      "packageUri": "lib/",
      "languageVersion": "3.1"
    },
    {
      "name": "fake_async",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/fake_async-1.3.1",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "flutter",
      "rootUri": "file:///C:/Users/andy/Documents/flutter/packages/flutter",
      "packageUri": "lib/",
      "languageVersion": "3.3"
    },
    {
      "name": "flutter_lints",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_lints-3.0.2",
      "packageUri": "lib/",
      "languageVersion": "3.1"
    },
    {
      "name": "flutter_test",
      "rootUri": "file:///C:/Users/andy/Documents/flutter/packages/flutter_test",
      "packageUri": "lib/",
      "languageVersion": "3.3"
    },
    {
      "name": "leak_tracker",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/leak_tracker-10.0.4",
      "packageUri": "lib/",
      "languageVersion": "3.2"
    },
    {
      "name": "leak_tracker_flutter_testing",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/leak_tracker_flutter_testing-3.0.3",
      "packageUri": "lib/",
      "languageVersion": "3.2"
    },
    {
      "name": "leak_tracker_testing",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/leak_tracker_testing-3.0.1",
      "packageUri": "lib/",
      "languageVersion": "3.2"
    },
    {
      "name": "lints",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/lints-3.0.0",
      "packageUri": "lib/",
      "languageVersion": "3.0"
    },
    {
      "name": "matcher",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/matcher-0.12.16+1",
      "packageUri": "lib/",
      "languageVersion": "3.0"
    },
    {
      "name": "material_color_utilities",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/material_color_utilities-0.8.0",
      "packageUri": "lib/",
      "languageVersion": "2.17"
    },
    {
      "name": "meta",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/meta-1.12.0",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "path",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/path-1.9.0",
      "packageUri": "lib/",
      "languageVersion": "3.0"
    },
    {
      "name": "sky_engine",
      "rootUri": "file:///C:/Users/andy/Documents/flutter/bin/cache/pkg/sky_engine",
      "packageUri": "lib/",
      "languageVersion": "3.2"
    },
    {
      "name": "source_span",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/source_span-1.10.0",
      "packageUri": "lib/",
      "languageVersion": "2.18"
    },
    {
      "name": "stack_trace",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/stack_trace-1.11.1",
      "packageUri": "lib/",
      "languageVersion": "2.18"
    },
    {
      "name": "stream_channel",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/stream_channel-2.1.2",
      "packageUri": "lib/",
      "languageVersion": "2.19"
    },
    {
      "name": "string_scanner",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/string_scanner-1.2.0",
      "packageUri": "lib/",
      "languageVersion": "2.18"
    },
    {
      "name": "term_glyph",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/term_glyph-1.2.1",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "test_api",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/test_api-0.7.0",
      "packageUri": "lib/",
      "languageVersion": "3.0"
    },
    {
      "name": "vector_math",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/vector_math-2.1.4",
      "packageUri": "lib/",
      "languageVersion": "2.14"
    },
    {
      "name": "vm_service",
      "rootUri": "file:///C:/Users/andy/AppData/Local/Pub/Cache/hosted/pub.dev/vm_service-14.2.1",
      "packageUri": "lib/",
      "languageVersion": "3.3"
    },
    {
      "name": "foobarapp4",
      "rootUri": "../",
      "packageUri": "lib/",
      "languageVersion": "3.4"
    }
  ],
  "generated": "2024-07-11T09:47:53.334815Z",
  "generator": "pub",
  "generatorVersion": "3.4.3",
  "flutterRoot": "file:///C:/Users/andy/Documents/flutter",
  "flutterVersion": "3.22.2",
  "pubCache": "file:///C:/Users/andy/AppData/Local/Pub/Cache"
}
```
It looks good, doesn't it? Nevertheless, `import 'package:flutter/material.dart';` in the main.dart file is underlined with a red color, obviously what follows has many errors as well (StatelessWidget etc.)

**Comment by andynewman10** [comment]
Are Dev Tools running the dart analyzer by launching a new process or are they doing it in-process? Maybe I can monitor everytime `dart` is run and see if it gets called when it should?

**Comment by DanTup** [comment]
> It looks good, doesn't it? Nevertheless, import 'package:flutter/material.dart'; in the main.dart file is underlined with a red color

Yes, if that `.dart_tool/package_config.json` was in the new project then it seems correct to me. I wonder if the project was written _before_ the `package_config` which resulted in those errors, then it didn't refresh properly when that file was written.

> Are Dev Tools running the dart analyzer by launching a new process or are they doing it in-process?

We run a single process `dart language_server` when you open the project and it runs in the background monitoring the files on disk.

Enabling the [instrumentation log](https://dartcode.org/docs/logging/#analyzer-instrumentation) will record all traffic between VS Code and the analysis server (as well as the file watch events the server reacts to). This might give some hints, but be aware that **it will also includes parts of files in your workspace** so may be difficult to share before you can repro this in a cleaner project.

One thing I would try is duplicating your project into a new folder and verify it still repros. Then try deleting _all_ of the `.dart` files, and then try again. If it still reproduces then, you may be much closer to a repro you can share?

**Comment by andynewman10** [comment]
> > It looks good, doesn't it? Nevertheless, import 'package:flutter/material.dart'; in the main.dart file is underlined with a red color
> 
> Yes, if that `.dart_tool/package_config.json` was in the new project then it seems correct to me. I wonder if the project was written _before_ the `package_config` which resulted in those errors, then it didn't refresh properly when that file was written.

main.dart:

LastModified=20240711114750.550595+120

package_config.json:

LastModified=20240711114753.347662+120

package_config.json has therefore been created after main.dart by the `flutter create` command when this command was ran.

**Comment by DanTup** [comment]
Yeah, that makes sense since probably the project is created and then `flutter pub get` is run to write the `package_config`. The analyzer _should_ handle this correctly though (and it sounds like it does when you do this in a new project), but maybe it's not doing it in some conditions (with your main project). If you can narrow it down to get a repro you can share, that would help me debug.

**Comment by andynewman10** [comment]
> Yeah, that makes sense since probably the project is created and then `flutter pub get` is run to write the `package_config`.

The `flutter create` command itself creates `.dart_tool/package_config.json`, independently from VS code.

**Comment by andynewman10** [comment]
I guess this is what you mean, my bad.

**Comment by DanTup** [comment]
Yeah, `flutter create` will do both. But the analyzer is just watching the disk, so it'll see the `pubspec.yaml` file created and then start building an analysis context for it, but this may happen before `package_config` is written. However, when `package_config` is written, it should recreate it and fix everything up.

If you're able to get steps to repro this that don't involve sharing your large project (it's possible duplicating the whole folder and deleting all/most of the .dart files will get you close?) that would really help.

**Comment by andynewman10** [comment]
This morning, I've got an error that I've never had before:
```
The Dart Tooling daemon has terminated

Source: Dart [Reload] [Show Log]
```
It occurs with both 3.96.0 and 3.98.0.
Is it serious? I know I have yet to provide a log or a simple repro for this issue, it is still on my TODO list.

**Comment by DanTup** [comment]
@andynewman10 could you file a new issue for it (it seems unrelated to this issue), and click the "Show Log" button to see if there is any information (like an exception or exit code) for the process.

**Comment by DanTup** [comment]
@andynewman10 thanks!

To answer your earlier question I missed:

> Is it serious?

The Dart Tooling Daemon (DTD) is a relatively new background process used for different Dart tools to communicate (for example the Flutter Sidebar that is hosted inside an iframe is part to DevTools and we use DTD to interact with it to tell it when devices change or debug sessions start/end). Right now, there's not a lot using it so it terminating is probably not fatal (though you may find the Flutter sidebar not working, for example). Over time more and more things will use it so if it's crashing, we'll definitely want to track down why.

**Comment by andynewman10** [comment]
@DanTup this bug is haunting me, and I've got another developer with a similar experience. I am desperately trying to find a solution.

Here is what has been done (today):

- completely uninstall VS code, using the Control Panel, with a reboot after the uninstall.
- delete `%USERPROFILE%\.vscode`
- delete the `Code` folder in `%APPDATA%`

once the above is done, Visual Studio Code is completely (100%) purged from the machine.

I did a `flutter pub clean` to be on the safe side, too.

Additionally, I deleted the flutter folder and got a fresh 3.24.5 version (known to work fine with this project).

Note that my machine is clean, it never crashes and I never had any data corruption.

Now, what I am doing is the following

1) from a shell, go to my Workspace folder
2) recursively perform a `flutter clean` from every single parent folder of all `.dart_tool` folders found. This makes sure everything is completely clean. I use a script for this.
3) make sure there is no `.vscode/settings.json` file.
3) launch VS code. Note that here, only the Flutter and Dart extensions are installed. I have no other extensions installed.
4) open my Workspace
5) the Dart Code extension proposes to do a `flutter pub get`: OK. This lasts a little while. From my shell, I see that `.dart_tool` folders are (indeed) progressively created.

In the VS `Output` log pane, here is what I see:

```
[myapplication\gs\gs] flutter pub get --no-example
Resolving dependencies...
Downloading packages...
  characters 1.3.0 (1.3.1 available)
  collection 1.18.0 (1.19.1 available)
  flutter_lints 2.0.3 (5.0.0 available)
  lints 2.1.1 (5.1.0 available)
  material_color_utilities 0.11.1 (0.12.0 available)
  meta 1.15.0 (1.16.0 available)
Got dependencies!
6 packages have newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.
C:\Users\andy\Documents\Workspace\myapplication\gs\gs\example\.dart_tool\package_config.json does not exist.
Did you run this command from the same directory as your pubspec.yaml file?

exit code 1
```
I am seeing this a few times while the Dart extension does its job. It seems that for a few packages/folders, this happens.

When checking on disk I see that `C:\Users\andy\Documents\Workspace\myapplication\gs\gs\example\.dart_tool\package_config.json` IS present.

It seems I get these errors for packages that I have a local copy of.

I can't provide the full workspace unfortunately, as it is too big. I could never narrow down the issue so far, despite trying really hard. Trimming the packages used in `myapplication` until I can pinpoint the problem would require a lot of time (disabling portions of the code scattered everywhere would be brutal...).

**Comment by andynewman10** [comment]
> Can you give more details about where the errors occur and which project they point to? It's not clear to me if by:
> 
> ```yaml
> foobarsubproject:
>   path: ./foobarsubproject
> ```
> 
> You mean you are adding a reference to "itself" here, or if you just mean you're using `path` dependencies between local projects. When this happens, if you open up the `.dart_tool/package_config.json` for the project where the analysis error is shown, does it contain a reference to the package that is "not found" and the correct path to it?
> 
> Does opening a file that has this error cause the error to go away?
> 
> Do you know when this issue started? Is it with a recent update of Dart/Flutter?

My folder structure is
```
myapplication/
  gs/
    gs/
      example/
        pubspec.yaml
      pubspec.yaml
  pubspec.yaml
```
myapplication's `.dart_tool/package_config.json` does list gs/gs:
```json
    {
      "name": "gs",
      "rootUri": "../gs/gs",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
```
`gs` is referenced this way in myapplication/pubspec.yaml:
```yaml
  gs:
    path: ./gs/gs
```
I am puzzled about the package_config.json error regarding `gs/gs/example/.dart_tool/package_config.json`. Would it be possible that, when this error shows up, the Dart extension code analysis is stopped?

**Comment by DanTup** [comment]
@andynewman10 

> @DanTup this bug is haunting me,

There are a couple of different things that were discussed above and it's not completely clear to me what the problem is you're having here. To keep things simple, could you open a new issue focused on just the specific issue you're seeing now? If you see diagnostics you don't expect, please include a screenshot if possible (even if it's small/mostly redacted).



> C:\Users\andy\Documents\Workspace\myapplication\gs\gs\example\.dart_tool\package_config.json does not exist.
> Did you run this command from the same directory as your pubspec.yaml file?

This error seems a little odd because the command shown has `--no-example`, though I'm not sure if it's related. Do any of your files (or pubspecs) outside of `example` reference the example project?


> Would it be possible that, when this error shows up, the Dart extension code analysis is stopped?

I don't think so, that error is coming from `pub` which is run as a separate process. However, it's not impossible that the reason this error occurs for pub also causes problems elsewhere, but there's not much to go on yet. Anything you can do make a stripped down repro you can share would definitely make it easier to track down.

**Comment by andynewman10** [comment]
> > C:\Users\andy\Documents\Workspace\myapplication\gs\gs\example.dart_tool\package_config.json does not exist.
> > Did you run this command from the same directory as your pubspec.yaml file?
> 
> This error seems a little odd because the command shown has `--no-example`, though I'm not sure if it's related. 

The `flutter pub get `command issued by Dart Code probably walks down the folder hierarchy in my workspace, and the command is probably issued using this `--no-example` flag (as I have always seen, including when the command is successful, when I get `exit code 0`)?

> Do any of your files (or pubspecs) outside of `example` reference the example project?

Definitely none of them reference this example project.

When I restart VS code, Code Analysis eventually finishes and the red color disappear.

But then, another symptom, which I can't get rid of, is the following, when trying to hover member variables when debugging, in some locations (typically, locations that match local copies of packages in subfolders):

![image](https://github.com/user-attachments/assets/38f15779-685a-40a9-a5e5-bf5326c331c9)

Above is a capture of my debugging session. I have a local copy of `google_maps_flutter_web`, and I have placed a breakpoint in one function. Hovering member variables only shows the type of the variable, not its contents as it should normally happen. I understand this might be another side effect of my original issue, but all I'm trying to do is provide relevant information.

**Comment by DanTup** [comment]
> When I restart VS code, Code Analysis eventually finishes and the red color disappear.

It's still unclear to me what the issue is here - is the issue diagnostics that are incorrect, or just that it's slow? (It would be good to open a new issue that clearly shows what the problem is, there are a lot of different things mentioned here that make it difficult to understand).


> Above is a capture of my debugging session. I have a local copy of google_maps_flutter_web, and I have placed a breakpoint in one function. Hovering member variables only shows the type of the variable

To avoid more confusion, we should open new issues for each problem. However in this case I think the issue is just that you're hovering over a variable before it exists - if you step to the next line and then hover it, I suspect you'll see the value.

**Comment by andynewman10** [comment]
The member variable contents cannot be displayed (I did the capture a bit too quickly):

![image](https://github.com/user-attachments/assets/562f255e-c101-4d6a-b3aa-d225c2761d25)

> It's still unclear to me what the issue is here - is the issue diagnostics that are incorrect, or just that it's slow?

It is not too slow.

> It would be good to open a new issue that clearly shows what the problem is, there are a lot of different things mentioned here that make it difficult to understand

I can certainly open another issue, but here we're dealing with the same issue and I understood it is better to gather all symptoms under the umbrella of the same issue. The current issue is:

- freshly opening a workspace on which flutter pub get has not yet been performed under VS results in flutter pub get issuing errors in the console and code analysis not being performed (even if you wait one hour)
- this results in red errors displayed everywhere in the source code on some computers. On these machines, even if you wait for a very long time, the errors will not disappear. In this case, only restarting Visual Studio will trigger full code analysis but other problems will show up (debug problem shown above).

Green light to open a new issue?

**Comment by andynewman10** [comment]
Adding a 'this' reference in the Watch pane results in this exception:

![image](https://github.com/user-attachments/assets/d784945a-67c5-43fb-83a5-6ea7e316924c)

**Comment by andynewman10** [comment]
I have managed to get rid of the problem. 

My workspace is organized in the following way:

```
- .vscode
- project1
- project2
- project3
... many other projects
```
One of the projects, say, projectN, has a Python virtual environment folder with lots of stuff in it, and is about 2GB in size in total with about 30,000 files. There is no Flutter/Dart/yaml content except for one file (port.yaml from a 'markdown_it' package). No Python extension is installed for my tests, making venv parsing issues unlikely.

When this folder is in my workspace folder tree, I never (ever) see the `Analyzing...` message in VS code, suggesting code analysis is not taking place as it should, even if I wait a full hour (I have a fast SSD M2 drive). Perhaps VS code gives up somehow. As a result, red errors remain scattered everywhere, and the only cure is to reload VS code - in this case, code analysis is indeed triggered, fortunately.

@DanTup outside deleting this folder - is there a way I can tell the Dart extension and/or VS code to never attempt to find anything useful in that folder? Some kind of '.vsignore' or '.dartignore'? Sorry to bother you again.
